### PR TITLE
[SPARK-38771][SQL] Adaptive Bloom filter Join

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR introduces an adaptive Bloom filter Join rule for `AQEOptimizer`.
It has two benefit:
1. It reduces spilling data when sorting.
2. It improves join performance, especially join keys may be skewed.

It only works:
1. AQE is enabled and can't be convert to BroadcastHashJoin.
2. The filtering side rows should be less than 100,000,000 rows.
3. The pruning side should be large and must spill data to disk when sorting(Larger than `spark.executor.memory(mb) * spark.memory.fraction * (1 - spark.memory.storageFraction) * 1024 * 1024 / spark.executor.cores  / factor * spark.sql.shuffle.partitions`). For example:
     Default config: `spark.executor.memory` = 1g,`spark.memory.fraction` = 0.6, `spark.memory.storageFraction` = 0.5, `spark.executor.cores` = 1 and `spark.sql.shuffle.partitions` = 200. The pruning side should larger than 30gb.
     Our production config: `spark.executor.memory` = 45g,`spark.memory.fraction` = 0.75, `spark.memory.storageFraction` = 0.5, `spark.executor.cores` = 18 and `spark.sql.shuffle.partitions` = 10000. The pruning side should larger than 4687gb.

How to disable this rule:
```
set spark.sql.adaptive.optimizer.excludedRules=org.apache.spark.sql.execution.adaptive.AdaptiveBloomFilterJoin
```

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test and benchmark test.
 Before | After
-- | --
image | image

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveBloomFilterJoin.scala [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    Filter(BloomFilterMightContain(bloomFilterSubquery, new XxHash64(pruningKeys)), pruningPlan)
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformDown {
+    case join @ ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, _, _, left, right, _)
+        if left.stats.isRuntime && right.stats.isRuntime && nonBroadcastHashJoin(join) =>
+      val leftRowCnt = left.stats.rowCount.get
+      val rightRowCnt = right.stats.rowCount.get
+
+      if (canPruneLeft(joinType) && muchSmaller(rightRowCnt, leftRowCnt) &&
```
Comment (weixiuli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveBloomFilterJoin.scala:99]: We should check the filterSide. rowCount  > 0 ?

Comment (singhpk234) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveBloomFilterJoin.scala:99]: will [AQEPropagateEmptyRelation](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEPropagateEmptyRelation.scala) not take care of empty side ? This rule runs before Adaptive BFJ rule ?

**Comment by wangyum** [comment]
Another case:

 Before | After
-- | --
image | image

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveBloomFilterJoin.scala [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    filterSide < maxNumItems && filterSide * 5 < pruningSide
+
+  private def avgSizePerPartition(logicalPlan: LogicalPlan): Float =
+    logicalPlan.stats.sizeInBytes.toFloat / conf.numShufflePartitions
+
+  private def nonBroadcastHashJoin(join: Join): Boolean = {
+    !canPlanAsBroadcastHashJoin(join, conf) && join.children.forall {
+      case LogicalQueryStage(_, stage: ShuffleQueryStageExec) => stage.isMaterialized
+      case _ => false
+    }
```
[the remark this answers is not in the document]
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveBloomFilterJoin.scala:71]: Yes.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveBloomFilterJoin.scala [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  private def canPruneRight(joinType: JoinType): Boolean = joinType match {
+    case Inner | LeftSemi | LeftOuter => true
+    case _ => false
+  }
+
+  private def muchSmaller(filterSide: BigInt, pruningSide: BigInt): Boolean =
+    filterSide < maxNumItems && filterSide * 5 < pruningSide
+
+  private def avgSizePerPartition(logicalPlan: LogicalPlan): Float =
+    logicalPlan.stats.sizeInBytes.toFloat / conf.numShufflePartitions
```
[the remark this answers is not in the document]
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveBloomFilterJoin.scala:51]: I think we should optimize Coalesce itself.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveBloomFilterJoin.scala [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  // The factor of raw data to Java objects.
+  private final val factor = 2
+
+  private val sc = sparkSession.sparkContext
+  private val fraction = sc.conf.get(config.MEMORY_FRACTION)
+  private val storageFraction = sc.conf.get(config.MEMORY_STORAGE_FRACTION)
+  private val executorCores = sc.conf.get(config.EXECUTOR_CORES)
+
+  private val memoryPerTask =
+    (sc.executorMemory * fraction * (1 - storageFraction) * (1L << 20)).toFloat / executorCores
```
[the remark this answers is not in the document]
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveBloomFilterJoin.scala]: I don't want to think about reserved memory.

**Comment by lsyldliu** [comment]
@wangyum  Does this pr have any progress?

**Comment by wangyum** [comment]
We have merged into our internal Spark. Here is an example:
image

**Pull request closed without merging**